sdk/python/feast/infra/offline_stores/bigquery.py

```diff
@@ -287,23 +291,36 @@ def to_arrow(self) -> pyarrow.Table:
 
 
 def block_until_done(client, bq_job):
-    def _is_done(job_id):
-        return client.get_job(job_id).state in ["PENDING", "RUNNING"]
+    """
+    Waits a maximum of 30 minutes for bq_job to finish running.
+    Args:
+        client: A bigquery.client.Client to monitor the bq_job.
+        bq_job: The bigquery.job.QueryJob that blocks until done runnning.
+    Raises:
+        BigQueryJobStillRunning exception if the function has blocked longer than 30 minutes.
+        BigQueryJobCancelled exception on a KeyboardInterrupt.
+    """
 
-    @retry(wait=wait_fixed(10), stop=stop_after_delay(1800), reraise=True)
+    @retry(
+        wait=wait_fixed(10),
+        stop=stop_after_delay(1800),
+        retry=retry_if_exception_type(BigQueryJobStillRunning),
+        reraise=True,
+    )
     def _wait_until_done(job_id):
-        return _is_done(job_id)
+        if client.get_job(job_id).state in ["PENDING", "RUNNING"]:
+            raise BigQueryJobStillRunning(job_id=job_id)
 
     job_id = bq_job.job_id
-    _wait_until_done(job_id=job_id)
+    try:
+        _wait_until_done(job_id=job_id)
+    except KeyboardInterrupt:
+        client.cancel_job(job_id)
+        raise BigQueryJobCancelled(job_id=job_id)
 
     if bq_job.exception():
         raise bq_job.exception()
 
-    if not _is_done(job_id):
```

Comment:
This piece of logic is still needed I think.

For example if the job timeout, an exception will be raised and the python script will end. But that won't cancel the BQ job.

Does that make sense ?

Reply:
By timeout do you mean it took longer than the 30 minutes we allowed it to retry for? If yes, I wanted to make the behavior such that the job will still continue in case it is a job that requires longer than 30 minutes. I could perhaps make the error message more descriptive - currently it would be:
```python
BigQueryJobStillRunning ->  "The BigQuery job with ID '{job_id}' is still running."
```

Another option would be to cancel any job running for longer than 30 minutes, then maybe asking the user to manually retry running their query outside of the Feast api, like:
```python
client.query(<BigQueryRetrievalJob>.to_sql(), job_config=job_config)
```
Downside is they have to wait at least another 30 minutes, unless you think the first timeout on the Feast side might not actually mean that the job was taking that long to run but instead some problem with the BQ client API?

Reply:
> Another option would be to cancel any job running for longer than 30 minutes, then maybe asking the user to manually retry running their query outside of the Feast api, like:

I was thinking that the solution would be to let the user pass a timeout if 30 min was not enough. I don't think that it's ideal to perform the query outside Feast.

Reply:
by `pass a timeout` do you mean like a flag or argument to increase the time limit?

Reply:
Yes

Something like the following if it works (never tried it)

```python
def block_until_done(client, bq_job, timeout=1800):
...
     @retry(wait=wait_fixed(10), stop=stop_after_delay(timeout), reraise=True)
```

Reply:
Sounds good! Looks like variables can't go in the decorator itself, so I'll attempt the second option [here](https://tenacity.readthedocs.io/en/latest/#changing-arguments-at-run-time):